### PR TITLE
Srch 3111 - update I14y to save the document's MIME type

### PR DESCRIPTION
**Pull request by jamesmadhoc**
## Summary
- Implemented as described in Acceptance Criteria and Dev approach, though with one decision not mentioned there: where to normalize the MIME type string:  I did this up in the [API] controller rather than down in the model, because the MIME::Types.include? method will accept as valid a valid "type/subtype" string followed by a semicolon (or whitespace -- basically a non-word character, I think) and then just about anything, and I thought we might conceivably want to use this behavior someday.
- Manual test behaved as described in Acceptance Criteria (once I did it correctly!).
 
### Checklist
Please ensure you have addressed all concerns below before marking a PR "ready for review" or before requesting a re-review. If you cannot complete an item below, replace the checkbox with the ⚠️ `:warning:` emoji and explain why the step was not completed.
 
#### Functionality Checks

- [x] You have run `bundle update` and committed your changes to Gemfile.lock.
 
- [x] You have merged the latest changes from the target branch (usually `main`) into your branch.
 
- [x] You have squashed your commits into a single commit (exceptions: your PR includes commits with formatting-only changes, such as required by Rubocop or Cookstyle, or if this is a feature branch that includes multiple commits).
 
- [x] Your primary commit message is of the format **SRCH-#### \<description\>** matching the associated Jira ticket.

- [x] PR title is either of the format **SRCH-#### \<description\>** matching the associated Jira ticket (i.e. "SRCH-123 implement feature X"), or **Release - SRCH-####, SRCH-####, SRCH-####** matching the Jira ticket numbers in the release.
 
- [x] Automated checks pass. If Code Climate checks do not pass, explain reason for failures: CC complains about identical hash literals at spec/models/document_spec.rb:10 and ...:18; leaving these as-is per reviewer's comments.
 
#### Process Checks

- [x] You have specified at least one "Reviewer".

### CODE REVIEW

File: spec/requests/api/v1/documents_spec.rb [hunk 1]
```diff
@@ -43,14 +43,17 @@
     end
 
     let(:valid_params) do
-      { document_id: id,
-        title:       'my title',
-        path:        'http://www.gov.gov/goo.html',
+      {
+        document_id: id,
```
Comment (MothOnMars) [inline spec/requests/api/v1/documents_spec.rb:47]: I can see that this is the spec that ran afoul of the `Layout/HashAlignment` cop. While I agree that the previous version is easier to read, I'm inclined to stick with the current configuration, which is the Rubocop default. There are a couple of issues with the "align both keys and values" style:
- it means that any change to the hash has the potential to require a re-alignment of all the other keys. Rubocop could do that for us automagically, but...
- the other downside is that it muddies the git history. It's harder to track the source and rational for changes when the hash has changed many times.
- most of the hashes in our various repos use the default format, so changing it would result in even more Rubocop offenses, across repos.

TL;DR: thanks for making these changes, even if you had to hold your nose.

**Review by MothOnMars — CHANGES_REQUESTED** [review]
See my suggestions. To clarify spec changes I mentioned, I pushed a branch with the suggested changes:
https://github.com/GSA/i14y/blob/2ba90eb0ff302e460dd5e7531cd2aeb272aa87d5/spec/models/document_spec.rb

File: Gemfile [hunk 2]
```diff
@@ -17,6 +17,8 @@ gem 'typhoeus', '~> 1.4.0'
 
 gem 'rake', '~> 13.0.0'
 
+gem 'mime-types', '~> 3.4'
```
[the remark this answers is not in the document]
Comment (jamesmadhoc) [reply Gemfile:20]: Removing mime-types in favor of mini_mime.

File: spec/models/document_spec.rb [hunk 3]
```diff
@@ -3,49 +3,72 @@
...
+  end
+
+  describe 'invalid?' do
+    subject(:document) { described_class.new(valid_params.merge(mime_type: 'text/not_a_valid_mime_type')) }
+
+    it 'returns true and sets an error message for invalid MIME type' do
+      expect(document.invalid?).to be true
+      expect(document.errors.messages[:mime_type]).to include 'is invalid'
+    end
   end
```
Comment (MothOnMars) [inline spec/models/document_spec.rb:67]: See my original comment (https://github.com/GSA/i14y/pull/200#discussion_r946938115) and the example spec I wrote (https://github.com/GSA/i14y/blob/2ba90eb0ff302e460dd5e7531cd2aeb272aa87d5/spec/models/document_spec.rb#L53). 

Please:
- move this into the `validations` block
- use a `context` rather than a `describe` block
- break the `subject` declaration into several lines for readability (the general rule of thumb is that lines should be no more than 80-100 char)

The error message spec should be separate from the validation spec (which can be a one-liner). We break this rule on occasion, but there's no need to combine them here. See https://www.betterspecs.org/#single

**Review by MothOnMars — CHANGES_REQUESTED** [review]

**Comment by jamesmadhoc** [comment]
Hi, Martha.  I think I've addressed your recent suggestions, so please re-review https://github.com/GSA/i14y/pull/200 .  Note that codeclimate shows two failures: identical hash literals at spec/models/document_spec.rb:10 and ...:28, but I've left them that way per your comments.

**Comment by MothOnMars** [comment]
@jamesmadhoc , I'll do a full re-review on Monday, but please disable the `duplication` engine for the `spec/` directory as we've done in search-gov:
https://github.com/GSA/search-gov/blob/main/.codeclimate.yml#L22

Duplicated code is fairly common and is often more readable than the alternative in specs.

**Review by MothOnMars — APPROVED** [review]
LGTM. Adding a "do not merge" label on this to ensure it's not merged before we add the mappings to the indices.

**Review by MothOnMars — CHANGES_REQUESTED** [review]
@jamesmadhoc , please make one more change to this PR: Add `mime_type` to the list of keyword fields:
```ruby
# app/templates/documents.rb

%w[document_id extension language mime_type path tags].each { |field| keyword(json, field) }
```
We don't currently have a spec for mappings, but I'm planning to add one before the next round of field additions. So for now, this is tested by existing integration specs that confirm it doesn't break anything.

For the moment, this change will only apply in dev/test environments, as it will not automatically update the I14y template or index mappings. Per a meeting I had w/ Amani yesterday, I'm going to update those in staging & production as a subtask of your story: https://cm-jira.usa.gov/browse/SRCH-3113

**Review by MothOnMars — DISMISSED** [review]
LGTM

File: app/controllers/api/v1/documents.rb [hunk 4]
```diff
@@ -1,5 +1,18 @@
...
 module Api
   module V1
+    class ValidMimeType < Grape::Validations::Base
+      def validate_param!(attr_name, params)
+        return if MiniMime.lookup_by_content_type(params[attr_name])
+
+        raise Grape::Exceptions::Validation.new(params: [@scope.full_name(attr_name)],
+                                                message: 'is not a valid MIME type')
+      end
+    end
```
Comment (MothOnMars) [inline app/controllers/api/v1/documents.rb:14]: @jamesmadhoc , this approach isn't ideal, because it duplicates the validation in the `Document` model. I suggest simplifying this by restricting the validation logic to the model. See https://github.com/MothOnMars/i14y/commit/749af63b714f080c691b17d7996f0911d0f7d497.

I14y is a little bit weird because the models and model-level validations were added very late in its life. This has led to what I now see are some inconsistencies and missing test cases for the controller. For now, please make the suggested changes, including the comment I've added re. the validations.

**Review by MothOnMars — CHANGES_REQUESTED** [review]

File: app/controllers/api/v1/documents.rb (hunk 4 above)
Comment (MothOnMars) [reply app/controllers/api/v1/documents.rb]: @jamesmadhoc , I just realized I linked to an older version of my commit. Please add this comment to guide future devs:
https://github.com/MothOnMars/i14y/commit/5c2f1c6e0#diff-04c8891d20866a2eeb884bcce73deff08c6c923b3676d06bcd32dad6d4da57baR9-R11

**Review by MothOnMars — CHANGES_REQUESTED** [review]
- add comment (https://github.com/MothOnMars/i14y/commit/5c2f1c6e0#diff-04c8891d20866a2eeb884bcce73deff08c6c923b3676d06bcd32dad6d4da57baR9-R11)
- resolve merge conflicts

**Review by MothOnMars — APPROVED** [review]
LGTM. The merge conflicts still need resolution.

**Comment by jamesmadhoc** [comment]
Backmerged from main yesterday, resolved conflicts, ran ``bundle update'', and got through CircleCI OK.

**Review by MothOnMars — APPROVED** [review]
LGTM

**Review by MothOnMars — APPROVED** [review]
LGTM
